File: apps/desktop/src/main/audio-host/wire/audio.ts
export interface AudioHostDevice {
  id: string
  name: string
  is_default: boolean
  default_sample_rate: number | null
  min_buffer_size: number | null
  max_buffer_size: number | null
  channel_count: number | null
}

export interface AudioHostApplicationCaptureLogicalTarget {
  platform: string
  bundle_identifier: string | null
  executable_path: string
  executable_name: string
  include_process_tree: boolean
}

export interface AudioHostApplicationCaptureTarget {
  runtime_id: string
  process_id: number
  display_name: string
  executable_path: string
  logical_target: AudioHostApplicationCaptureLogicalTarget
  channel_count: number
  status: string
}

export interface AudioHostApplicationCaptureSnapshot {
  runtime_id: string
  process_id: number | null
  display_name: string
  executable_path: string
  logical_target: AudioHostApplicationCaptureLogicalTarget
  channel_count: number
  status: string
  dropout_frames: number
  overflow_frames: number
  underflow_frames: number
}

export interface AudioHostRuntime {
  state: string
  requested_buffer_size: number | null
  sample_rate: number | null
  input_sample_rate: number | null
  output_sample_rate: number | null
  input_buffer_size: number | null
  output_buffer_size: number | null
  ring_buffer_capacity_frames: number | null
  ring_buffer_fill_frames: number | null
  input_latency_ms: number | null
  output_latency_ms: number | null
  ring_buffer_latency_ms: number | null
  engine_latency_ms: number | null
  estimated_round_trip_latency_ms: number | null
  xruns: number
  clock_sync: string
  buffer_fallback: boolean
}

export interface AudioHostDeviceRecovery {
  recovery_id: number
  revision: number
  candidate_revision: number
  attempt_generation: number
  phase:
    | "waiting-for-authorization"
    | "waiting-for-change"
    | "attempting-original"
    | "original-restored"
    | "applying-selection"
    | "selection-failed"
  original_config: {
    backend: string
    input_device_id: string
    output_device_id: string
    buffer_size: number
    session_sample_rate: number | null
  }
  candidates: { inputs: AudioHostDevice[]; outputs: AudioHostDevice[] }
  lost_directions: Array<"input" | "output">
  fault:
    | "device-not-available"
    | "stream-invalidated"
    | "host-unavailable"
    | "device-busy"
    | "backend-error"
}

export interface AudioHostRoundTripLatencyMeasurement {
  status: string
  input_channel: number | null
  output_channel: number | null
  measured_round_trip_latency_ms: number | null
  failure: string | null
}

export interface AudioHostMeter {
  channel_id: string
  pre_left: number
  pre_right: number
  post_left: number
  post_right: number
  held_left: number
  held_right: number
  clipped: boolean
}

export interface AudioHostBenchmarkReport {
  duration_ms: number
  overall_realtime_factor: number
  worst_p99_deadline_utilization_percent: number
  scenarios: Array<{
    id: string
    label: string
    description: string
    sample_rate: number
    block_size: number
    tracks: number
    buses: number
    sends: number
    plugins: number
    elapsed_ms: number
    audio_duration_ms: number
    average_block_ms: number
    p95_block_ms: number
    p99_block_ms: number
    max_block_ms: number
    buffer_budget_ms: number
    p99_deadline_utilization_percent: number
    deadline_misses: number
    measured_blocks: number
    realtime_factor: number
  }>
}

export interface AudioHostTransport {
  state: string
  position_frames: number
  position_ticks: number
  sample_rate: number
  effective_bpm: number | null
  clock_source: string
  waiting_for: string | null
  loop_enabled: boolean
  loop_start_tick: number | null
  loop_end_tick: number | null
}

File: crates/dsp-runtime/src/protocol/audio.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AudioEngineConfig {
    pub backend: String,
    pub input_device_id: String,
    pub output_device_id: String,
    pub buffer_size: u32,
    pub session_sample_rate: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AudioBackend {
    pub id: String,
    pub label: String,
    pub available: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AudioDevice {
    pub id: String,
    pub name: String,
    pub is_default: bool,
    pub default_sample_rate: Option<u32>,
    pub min_buffer_size: Option<u32>,
    pub max_buffer_size: Option<u32>,
    pub channel_count: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AudioDeviceList {
    pub inputs: Vec<AudioDevice>,
    pub outputs: Vec<AudioDevice>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum AudioStreamDirection {
    Input,
    Output,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum AudioDeviceFaultKind {
    DeviceNotAvailable,
    StreamInvalidated,
    HostUnavailable,
    DeviceBusy,
    BackendError,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum AudioDeviceRecoveryPhase {
    WaitingForAuthorization,
    WaitingForChange,
    AttemptingOriginal,
    OriginalRestored,
    ApplyingSelection,
    SelectionFailed,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AudioDeviceRecovery {
    pub recovery_id: u64,
    pub revision: u64,
    pub candidate_revision: u64,
    pub attempt_generation: u64,
    pub phase: AudioDeviceRecoveryPhase,
    pub original_config: AudioEngineConfig,
    pub candidates: AudioDeviceList,
    pub lost_directions: Vec<AudioStreamDirection>,
    pub fault: AudioDeviceFaultKind,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AudioRuntime {
    pub state: String,
    pub requested_buffer_size: Option<u32>,
    pub sample_rate: Option<u32>,
    pub input_sample_rate: Option<u32>,
    pub output_sample_rate: Option<u32>,
    pub input_buffer_size: Option<u32>,
    pub output_buffer_size: Option<u32>,
    pub ring_buffer_capacity_frames: Option<u32>,
    pub ring_buffer_fill_frames: Option<u32>,
    pub input_latency_ms: Option<f64>,
    pub output_latency_ms: Option<f64>,
    pub ring_buffer_latency_ms: Option<f64>,
    pub engine_latency_ms: Option<f64>,
    pub estimated_round_trip_latency_ms: Option<f64>,
    pub xruns: u32,
    pub clock_sync: String,
    pub buffer_fallback: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ApplicationCaptureLogicalTarget {
    pub platform: String,
    #[serde(default)]
    pub bundle_identifier: Option<String>,
    pub executable_path: String,
    pub executable_name: String,
    pub include_process_tree: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ApplicationCaptureTargetDescriptor {
    pub runtime_id: String,
    pub process_id: u32,
    pub display_name: String,
    pub executable_path: String,
    pub logical_target: ApplicationCaptureLogicalTarget,
    pub channel_count: u32,
    pub status: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ApplicationCaptureSnapshot {
    pub runtime_id: String,
    pub process_id: Option<u32>,
    pub display_name: String,
    pub executable_path: String,
    pub logical_target: ApplicationCaptureLogicalTarget,
    pub channel_count: u32,
    pub status: String,
    pub dropout_frames: u64,
    pub overflow_frames: u64,
    pub underflow_frames: u64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct RoundTripLatencyMeasurementRequest {
    pub input_channel: u32,
    pub output_channel: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RoundTripLatencyMeasurement {
    pub status: String,
    pub input_channel: Option<u32>,
    pub output_channel: Option<u32>,
    pub measured_round_trip_latency_ms: Option<f64>,
    pub failure: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AudioBenchmarkScenario {
    pub id: String,
    pub label: String,
    pub description: String,
    pub sample_rate: u32,
    pub block_size: u32,
    pub tracks: u32,
    pub buses: u32,
    pub sends: u32,
    pub plugins: u32,
    pub elapsed_ms: f64,
    pub audio_duration_ms: f64,
    pub average_block_ms: f64,
    pub p95_block_ms: f64,
    pub p99_block_ms: f64,
    pub max_block_ms: f64,
    pub buffer_budget_ms: f64,
    pub p99_deadline_utilization_percent: f64,
    pub deadline_misses: u32,
    pub measured_blocks: u32,
    pub realtime_factor: f64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AudioBenchmarkReport {
    pub duration_ms: f64,
    pub overall_realtime_factor: f64,
    pub worst_p99_deadline_utilization_percent: f64,
    pub scenarios: Vec<AudioBenchmarkScenario>,
}
